### ISSUE TITLE
Files search takes a long time with unified search

### ISSUE THREAD

**Issue by jancborchardt** [nextcloud/server#23835]
Depending on the amount of files, the file search component of unified search takes a long time to get results back. It’s the slowest of all – Talk, Mail and everything loads before. On our own instance, it takes roughly 40 seconds.

Also, there is no x in the input field during loading so it’s impossible to cancel the search – you have to wait for it to finish to start a different one. To fix this, we could replace the spinner with an x on hover/focus.

Gif for illustration, showing how it takes 40 seconds:
![Search takes 40 seconds](https://user-images.githubusercontent.com/925062/97861814-a62d6680-1d04-11eb-9adc-f94f12a24070.gif)

(This is very related to the direct filtering of the current folder, which also doesn’t work: https://github.com/nextcloud/server/issues/23432 )

cc @nextcloud/search

**Comment by skjnldsv** [comment]
> Also, there is no x in the input field during loading so it’s impossible to cancel the search – you have to wait for it to finish to start a different one. To fix this, we could replace the spinner with an x on hover/focus.

Searching cancel other searches.
You can always change your query and it should trigger the search again

**Comment by ChristophWurst** [comment]
It's mainly slow because the files search implementation doesn't paginate. So you'd always build the full results list on the back-end before it's chopped into the individual pages. This doesn't scale, as you observed. *Our* instance just has a lot of files and shares, hence the slowness.

**Comment by andreas-p** [comment]
@ChristophWurst non-pagination is NOT the problem:

Pasting `152151` into the search term will yield a single result after ~4.7s (runtime taken from Firefox devtools).

Executing a query directly on the PostgreSQL 12.0 server (2.6e6 rows in oc_filecache):
- `SELECT * FROM oc_filecache WHERE name LIKE '%152151%`: 0.5s
- `SELECT * FROM oc_filecache WHERE name ILIKE '%152151%`: 1.1s

This makes NC file search at least 4 times slower than the database full table search.

If the search term is not entered by copy/paste, but typed in, things get much worse, since each character will trigger a new query on the server, consuming more resources (when typing fast, the final search request will take 16s).

@skjnldsv A new search does not cancel the running queries in the database backend.

**Comment by skjnldsv** [comment]
> @skjnldsv A new search does not cancel the running queries in the database backend.

Right, it cancel the frontend request of course.

**Comment by andreas-p** [comment]
I enabled log_min_duration_statement=800 in the PostgreSQL Server, and found that each search will execute the same query four times, which explains the x4 observed above.

**Comment by ChristophWurst** [comment]
> @ChristophWurst non-pagination is NOT the problem:

If there was pagination the query would include more `WHERE` clauses to scope the range. It would therefore help.

But yeah, `ILIKE` is expensive.

**Comment by andreas-p** [comment]
Not really, if the result set is small, or not at all when using window functions, since the time consuming part is the sequential scan, not the result set transfer.

For PostgreSQL, speed could be improved by not using `name ILIKE :searchExpression` (1.1s), but `lower(name) ILIKE lower(:searchExpression)` (0.8s) or even better `lowercasedname LIKE lower(:searchExpression)` (0.5s).

But there are fruits hanging much lower:
- eliminating those 3 exceeding queries, there's something going awfully wrong here.
- delaying search for some 300-500ms after the last keystroke, to prevent multiple zombie queries eating database performance.

More advanced:
- split filesearch in two requests: First search the current directory only (my example has 4800 Files in the folder, <20ms), then the full scan. Quite often, only the current directory is of interest.

**Comment by andreas-p** [comment]
I found the reason why the database was hammered with the same query 4x for a single search, see https://github.com/nextcloud/server/pull/24021

**Comment by ohthehugemanatee** [comment]
Suggest re-titling this issue in line with #24029 , since this *is* a regression in behavior from NC19. At best, we have lost the "fast search" capability. At worst, we've lost "search" alltogether.
 
My personal use case: updated to NC20 last night. Today, any search query will peg the mysql server and make nextcloud unresponsive. Having just performed a `files:clean` my `oc_filecache` table has 2057547 rows. It takes 2 seconds just to run the count query. :|

**Comment by icewind1991** [comment]
https://github.com/nextcloud/server/pull/25136 should help with search performance, how much of a difference it makes will depend greatly on setup specifics.

**Comment by ClCfe** [comment]
```WHERE`  (`name`  COLLATE utf8mb4_general_ci LIKE '%testing%') ```
Doesn't use column index and do a full table scan

What I would do :
Get rid of the collate (convert the column) and add  a "reverse_name" column with index (reverse_name=reverse(name))

Then query like this
```WHERE`  (`name` LIKE 'testing%' OR `reverse_name` LIKE 'gnitset%') ```

I know it is ugly and uses more storage, but it works

**Comment by jospoortvliet** [comment]
A quick search just in the current folder, like the old search this - basically adding a search provider that is named 'current folder' would do the trick to at least give some results quickly, covering many cases, right? what do you think @jancborchardt

**Comment by putt1ck** [comment]
I think there's something more; if people get frustrated by getting no result they change the search, but the old search continues to put load on the DB. We saw a load of 60 on the DB server just from a few users trying to use search and getting frustrated with it. Why would the search continue if the user moves on (to another search or manually looking for the files)?

**Comment by wiswedel** [comment]
> Why would the search continue if the user moves on (to another search or manually looking for the files)?

Good point. That then even applies to a search on the users and apps pages: Search for a user or an app, find quickly what you've been looking for (since those pages instantly narrow down), but have the unrelated unified search load remaining in the background.

**Comment by putt1ck** [comment]
We're having increased frustration levels with the search issue (big instance as in many files); is there anything we can do to patch this, either to improve performance back to NC19 levels or limit search to specific groups -  even disable, unless that prevents user lookups for sharing.

**Comment by ChristophWurst** [comment]
There were optimizations as far as I know, but most of the technical bottlenecks are yet to be improved.

**Comment by ChristophWurst** [comment]
Some improvements fresh out of the oven: https://github.com/nextcloud/server/pull/26049

**Comment by andreas-p** [comment]
Well that improvement is merely drop in the ocean. It limits the amount of data being transferred (if not cancelled anyway by the next longer search term), but will not tackle the multitude of huge database queries.

**Comment by ChristophWurst** [comment]
Well, I tried to help and keep you updated.

**Comment by andreas-p** [comment]
@ChristophWurst updates are appreciated.
Are there ongoing efforts to fix the issue fundamentally, i.e. provide a file search that works and unified search that won't DoS the db server?

**Comment by ChristophWurst** [comment]
I can't tell because I don't work on those internals.

**Comment by iwrbr** [comment]
Because this unified-search does indeed DoS-es the db server when searching files among lots and lots of files in the filecache table (searching for files was my target), and due to the absence of a real hope in fixing this issue in near future releases of NC, I decided to investigate the matter and I found a solution to avoid hanging the db server in the process of searching files. 
This solution implies changes in database as well in several files, so make sure you are not endangering your data.

I worked on this test instance:
NC: 20.0.6 
Operating System: CentOS 7.5
CPU: Intel(R) Xeon(R) CPU E5-2630 @ 2.30GHz (6 cores)
Memory: 16 GB
Filecache table: 1,364,426 rows (262MB size)

Searching for the string "Endpoint.msi" took:
- before: 233s
- after: 8s

Modifications I made:

First I've created an index called **fs_storage** in the table **oc_filecache**: 

`ALTER TABLE test.oc_filecache ADD INDEX 'fs_storage'('storage') USING BTREE;`


Then I modified four php functions to use this index, as follows:

**/lib/private/DB/QueryBuilder/QueryBuilder.php**
![image](https://user-images.githubusercontent.com/43100898/112537631-9861a600-8db7-11eb-8334-f474efead1b5.png)

**/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Query/QueryBuilder.php**
![image](https://user-images.githubusercontent.com/43100898/112537700-aca5a300-8db7-11eb-8a2f-0713584e609c.png)

**/lib/private/Files/Cache/CacheQueryBuilder.php**
![image](https://user-images.githubusercontent.com/43100898/112537665-a1eb0e00-8db7-11eb-8f56-0c2a55950e9b.png)

**/lib/private/Files/Cache/Cache.php**
![image](https://user-images.githubusercontent.com/43100898/112537749-b7f8ce80-8db7-11eb-9b91-dea610cb960a.png)




The problem is not solved entirely, the developers will most likely come to an elegant and more complex solution to fix this issue someday (hope sooner than later), but, for the time being one will not render the database unusable when searching for a file in a plethora of rows.

**Comment by wiswedel** [comment]
Thanks for your great contribution, @iwrbr 👏

@icewind1991 mind giving your opinion?

**Comment by ChristophWurst** [comment]
@iwrbr nice findings. Do you know why the index has to be forced and the db doesn't pick it up itself? The changes you made to doctrine won't be trivial to get upstreamed and they also contain db-specific SQL syntax that won't work on SQLite for example.

**Comment by andreas-p** [comment]
Interesting this fix works on some situations, but it's very mysql (?) and situation specific:
- PostgreSQL won't allow forcing an index. If the query planner decides there's a matching index and it's worth using it (i.e. the query provides some selectivity) it will be picked up automatically. 
- There are already indexes on oc_filecache that have storage as first column. Any db server should be able to use it partially.
- In my installation, most files are owned by one common user, sharing with all. Thus, the storage column has a very bad selectivity, and using the index would increase the total query cost.
- The proposed fix shows nicely why the postgresql community refuses to implement query hints and FORCE INDEX since the beginning: you can't know the cost of a query upfront for all different data distributions, and what's right in one situation might be very wrong in the next situation.

Conclusion: Trying to tune the database won't lead anywhere for this issue.

**Comment by iwrbr** [comment]
Without forcing the **fs_storage** index (or instructing the database to use that index bt using `USE INDEX (fs_storage)` instead of `FORCE INDEX (fs_storage)`) it uses the **fs_storage_path_hash** as primary index which does not help in this activity, hence only the creation of **fs_storage** is not enough. 
On the other side, using this **fs_storage** index as primary index on any other activities (except searching) that require file cache, delays considerably the execution. This is the reason I "forced" the usage of that index only when searching for files.

**Running explain on the query (without forcing the index):**
![image](https://user-images.githubusercontent.com/43100898/112648111-d7443a00-8e51-11eb-8764-17bbc20ff3f9.png)

It is true I worked on a MySQL database (because that is what I use), and I didn't say it was a permanent fix for all distributions and all db platforms. All I wanted was to show a fix that I used to partially solve the problem _for me_.

Waiting for the developers to solve the problem in its complexity, if one can use the fix and solve one's problem, then one may use it, if not, then don't.

The unified search lacks more than the search speed, the four most important features it lacks (imho) being:
- the ability to cancel search during search (by use of a button)
- the ability to delay the start of searching when typing letters in the search field (typing a 5 letter word triggers a minimum of 2 db queries, 7 letter word triggers min 3 db queries, no matter how fast I type - and I can type pretty fast; the only moment a single query is triggered is when using CUT+PASTE with the search string)
- the ability to cancel the running queries in the database backend when new search occurs, as @andreas-p stated above
- the ability to select what/where to search (i.e. search only in comments, or only for files, etc)
 
PS FYI: I just tried my fix (my MySQL fix) on the same server, but with all 900k+ files and folders moved in a group folder (instead of being divided among users) and the results are the same.

**Comment by wiswedel** [comment]
> * the ability to cancel search during search (by use of a button)

including the most obvious button _esc_ please!

**Comment by andreas-p** [comment]
@iwrbr and @wiswedel as mentioned above, cancel won't stop the execution of the query in the database backend, so DoS is still happening. 
It's mandatory to prevent firing those queries up front. As soon as the're issued to the backend, the database server will work its long hard way until it has the result. Cancelling would prevent retrieving the result set, which is just a (usually neclectable) fraction of the total cost.

**Comment by wiswedel** [comment]
> @iwrbr and @wiswedel as mentioned above, cancel won't stop the execution of the query in the database backend, so DoS is still happening.
> It's mandatory to prevent firing those queries up front. As soon as the're issued to the backend, the database server will work its long hard way until it has the result. Cancelling would prevent retrieving the result set, which is just a (usually neclectable) fraction of the total cost.

I guess @iwrbr did assume actual termination in his requirement list (which I'd back 100%). But thanks for concretizing, @andreas-p 👍

**Comment by maiolif** [comment]
> NB is search much faster in NC21?

IMHO nothing has changed; webdav serach is also faster. But search still remain very slow

**Comment by glorenzutti** [comment]
I am having a problem similar to the one reported here. After having updated the nextcloud to its latest version, 21.0.1

File and document searches are very slow, to the point that I never get a result and the database server runs out of available connections.

When a search is made, the query that is made is the following:

 SELECT `file``fileid`,` storage`, `path`,` path_hash`, `file``parent`,` name`, `mimetype`,` mimepart`, `size`,` mtime`, ` storage_mtime`, `encrypted`,` etag`, `permissions`,` checksum`, `metadata_etag`,` creation_time`, `upload_time` FROM` oc_filecache` `file` LEFT JOIN` oc_filecache_extended` `fe` ON` file`. `fileid` =` fe``fileid` WHERE (`storage` = 1610) AND (` name` COLLATE utf8mb4_general_ci LIKE '% vpn%') ORDER BY `mtime` desc LIMIT 5

That query stays forever, and you never get a result.

If I take out the where storage ... the query is immediate. With which I deduce that the problem is some index. But adding it is not enough.

I also found that the oc_filecache_extended table, which I don't quite understand what it is for, if I delete it completely, the search is instantaneous again. But with at least few registrations, the search engine stops working.

mysql> select * from oc_filecache_extended;
+ ---------- + --------------- + --------------- + ------ ------- +
| fileid | metadata_etag | creation_time | upload_time |
+ ---------- + --------------- + --------------- + ------ ------- +
| 92906168 | NULL | 0 | 1619623232 |
| 92906196 | NULL | 0 | 1619623359 |
| 92906200 | NULL | 0 | 1619623579 |
| 92906201 | NULL | 0 | 1619623579 |
| 92906236 | NULL | 0 | 1619623697 |
| 92906260 | NULL | 0 | 1619623777 |
| 92906299 | NULL | 0 | 1619623923 |
+ ---------- + --------------- + --------------- + ------ ------- +
7 rows in set (0.00 sec)

With only those records, the search engine query does NOT work anymore.

I tried the solution proposed by @iwbr, but nothing changed.

It doesn't seem reasonable to me to be deleting the contents of that table all the time. I did not find documentation for what it is for ... But if the table oc_filecache_extended has at least ONE record, the search engine no longer works.

For now, I'm all the time deleting the contents of that table ... But it's crazy.

Does anyone have any idea what it could be due to?

**Comment by gentakojima** [comment]
Thank you for your help, specifically @iwrbr and @glorenzutti. 

I've played around creating indexes for storage, name and path but without any good results. @iwbr optimizations seems to make this problem better, but it's still very slow for me. I don't feel like deleting the contents on `oc_filecache_extended` without further knowledge of what I am deleting, though.

I've ended up disabling search by returning a null at the start of Cache::searchQuery method (around line 841 of `lib/private/Files/Cache/Cache.php`) . Hoping not to break anything else, but at least the search queries return now empty as fast as possible and the SQL queries are not there running forever.

**Comment by glorenzutti** [comment]
I couldn't find documentation regarding table oc_filecache_extended.

I did check the trashbin, versions, sharing, everything works with that table empty. I dunno what it is for.

Im aware this is not a good option. But all my users are demanding to be able to search.

I am testing the fulltextsearch plugin. But I don't see it usable, the first indexing is impossible when you have 40 TB of information.

**Comment by putt1ck** [comment]
Hard to say what the cause is, not seeing a lot of posts from the unified search devs. Does your instance have a lot of files? Our tests suggested that the search works ok if there's not many files (not many being thousands rather than hundreds of thousands)

**Comment by glorenzutti** [comment]
The size of my oc_filecache table is 30 GB. 
Im fully aware that the table never shrinks, and that I have even remove some external shares and nothing changes.
But I know that this is a known problem.
Im not sure how many files I have between all of my fileservers. The nextcloud says almost 8 millions.

**Comment by gentakojima** [comment]
> Im not sure how many files I have between all of my fileservers. The nextcloud says almost 8 millions.

Can you post the number of rows of your oc_filecache table? That's 10 millions for me.

```
> select count(*) from oc_filecache;
+----------+
| count(*) |
+----------+
| 10784280 |
+----------+
```

**Comment by glorenzutti** [comment]
I will update you with this in a few hours. I dont have shell right now.

**Comment by glorenzutti** [comment]
mysql> select count(*) from oc_filecache;
+----------+
| count(*) |
+----------+
|  7989103 |
+----------+

**Comment by iwrbr** [comment]
I just discovered that applying some modifications on the SQL query, and whitout the use of FORCE INDEX (to make @andreas-p happy again :) ), the response time plummeted from 200s to 1.5s, but let me take it step by step.

Prerequisites:
NC: 20.0.6
Operating System: CentOS 7.5
CPU: Intel(R) Xeon(R) CPU E5-2630 @ 2.30GHz (6 cores)
Memory: 16 GB
Filecache table: 1,373,295 rows (288MB size)
Filecache_extended table: 810,000 rows (22MB size)

The values `1` for `storage` and `err` for `name` are particular for my case (I searched for the `err` string and my storage has the value of 1).

1. I added an index to the filecache table as follows

`ALTER TABLE test.oc_filecache ADD INDEX 'fs_storage_name'('storage', 'name') USING BTREE;`

2. The original SQL query is (plain original, without FORCE INDEX or any modifications I made in the previous [comment](https://github.com/nextcloud/server/issues/23835#issuecomment-807416207)):

SELECT `filecache`.`fileid`, `storage`, `path`, `path_hash`, `filecache`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` 
FROM `oc_filecache` `filecache` 
LEFT JOIN `oc_filecache_extended` `fe` ON `filecache`.`fileid` = `fe`.`fileid` 
WHERE 
(`storage` = 1) 
AND 
(`name` COLLATE utf8mb4_general_ci LIKE '%err%')

**QUERY RESPONSE TIME: 200s** 

3. I modified the SQL query as follows:

SELECT `filecache`.`fileid`, `storage`, `path`, `path_hash`, `filecache`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time`
FROM `oc_filecache` `filecache`
LEFT JOIN `oc_filecache_extended` `fe` ON `filecache`.`fileid` = `fe`.`fileid`
WHERE `filecache`.`fileid` IN (
SELECT `fileid` FROM `oc_filecache`
WHERE 
(`storage` = 1) 
AND
(`name` COLLATE utf8mb4_general_ci LIKE '%err%')
)

**QUERY RESPONSE TIME: 1.5s** 


So, the only 2 modifications I made were to add an index (without this index nothing happens) and to change the WHERE clause a bit and boom, 1.5s.

I will try to modify the `SelectFileCache` function in `/lib/private/Files/Cache/CacheQueryBuilder.php` to match the query at step 3, but it may take some time and I make no promises I will succeed. If one of you succeeds faster than me, then please do.

**Comment by glorenzutti** [comment]
I'm going to cry. you're clearly on the road ...


mysql> SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM `oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE (`storage` = 3679) AND (`name`  COLLATE utf8mb4_general_ci LIKE '%gamboa r%') ORDER BY `mtime` desc LIMIT 5;
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
| fileid  | storage | path                                                                                                                                           | path_hash                        | parent  | name                                    | mimetype | mimepart | size   | mtime      | storage_mtime | encrypted | etag          | permissions | checksum | metadata_etag | creation_time | upload_time |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
| 1985339 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/ARCHIVO DIGITAL DAT/2015/4100 A 4299/4234 CLAROS GAMBOA RICHARD -SA                                        | 32185beab9c1ebfe1e91a98414473147 | 1971399 | 4234 CLAROS GAMBOA RICHARD -SA          |        2 |        1 | 723592 | 1619656474 |    1449166154 |         0 | 6089ffbc57249 |          31 |          | NULL          |          NULL |        NULL |
| 1984190 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/ARCHIVO DIGITAL DAT/2015/4500 A 4699/4680 GAMBOA ROMERO MARIANA LUCERITO- SA                               | d0bf5b21f0ada9c8a98b2812868803a8 | 1971410 | 4680 GAMBOA ROMERO MARIANA LUCERITO- SA |        2 |        1 | 935021 | 1619656466 |    1507829276 |         0 | 6089ffbb39b8a |          31 |          | NULL          |          NULL |        NULL |
| 1880772 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/REGISTRO UNICO DE ASISTIDOS/3487 GAMBOA ROMERO Mariana Lucerito                    | 568d1c266c64d7a0b84affba1dd29ded | 1740556 | 3487 GAMBOA ROMERO Mariana Lucerito     |        2 |        1 | 410283 | 1619655468 |    1508426544 |         0 | 6089ffbc57249 |          31 |          | NULL          |          NULL |        NULL |
| 1779153 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/Informes Nube/Maria/Terminados y entregados/Estadistica/pasados/Gamboa Romero.docx | 421d8bbf34acb3ba92f6c05d34795c34 | 1779109 | Gamboa Romero.docx                      |       23 |        3 | 381399 | 1524237497 |    1524237497 |         0 | 6089f7e482b6a |          27 |          | NULL          |          NULL |        NULL |
| 1776759 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/Maria/Terminados y entregados/Estadistica/pasados/Gamboa Romero.docx               | 8faeb36ae7f3191e305b4c3f1889d2dc | 1776738 | Gamboa Romero.docx                      |       23 |        3 | 381399 | 1524237497 |    1524237497 |         0 | 6089f7d58b9bc |          27 |          | NULL          |          NULL |        NULL |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
**5 rows in set (6.51 sec)**

mysql> SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM `oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE `file`.`fileid` IN ( select fileid FROM oc_filecache WHERE (`storage` = 3679) AND (`name`  COLLATE utf8mb4_general_ci LIKE '%gamboa r%') ) ORDER BY `mtime` desc LIMIT 5;
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
| fileid  | storage | path                                                                                                                                           | path_hash                        | parent  | name                                    | mimetype | mimepart | size   | mtime      | storage_mtime | encrypted | etag          | permissions | checksum | metadata_etag | creation_time | upload_time |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
| 1985339 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/ARCHIVO DIGITAL DAT/2015/4100 A 4299/4234 CLAROS GAMBOA RICHARD -SA                                        | 32185beab9c1ebfe1e91a98414473147 | 1971399 | 4234 CLAROS GAMBOA RICHARD -SA          |        2 |        1 | 723592 | 1619656474 |    1449166154 |         0 | 6089ffbc57249 |          31 |          | NULL          |          NULL |        NULL |
| 1984190 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/ARCHIVO DIGITAL DAT/2015/4500 A 4699/4680 GAMBOA ROMERO MARIANA LUCERITO- SA                               | d0bf5b21f0ada9c8a98b2812868803a8 | 1971410 | 4680 GAMBOA ROMERO MARIANA LUCERITO- SA |        2 |        1 | 935021 | 1619656466 |    1507829276 |         0 | 6089ffbb39b8a |          31 |          | NULL          |          NULL |        NULL |
| 1880772 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/REGISTRO UNICO DE ASISTIDOS/3487 GAMBOA ROMERO Mariana Lucerito                    | 568d1c266c64d7a0b84affba1dd29ded | 1740556 | 3487 GAMBOA ROMERO Mariana Lucerito     |        2 |        1 | 410283 | 1619655468 |    1508426544 |         0 | 6089ffbc57249 |          31 |          | NULL          |          NULL |        NULL |
| 1779153 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/Informes Nube/Maria/Terminados y entregados/Estadistica/pasados/Gamboa Romero.docx | 421d8bbf34acb3ba92f6c05d34795c34 | 1779109 | Gamboa Romero.docx                      |       23 |        3 | 381399 | 1524237497 |    1524237497 |         0 | 6089f7e482b6a |          27 |          | NULL          |          NULL |        NULL |
| 1776759 |    3679 | Direccion-de-Asistencia-Tecnica-MPD/DAT 2018/Trabajo Social/Maria/Terminados y entregados/Estadistica/pasados/Gamboa Romero.docx               | 8faeb36ae7f3191e305b4c3f1889d2dc | 1776738 | Gamboa Romero.docx                      |       23 |        3 | 381399 | 1524237497 |    1524237497 |         0 | 6089f7d58b9bc |          27 |          | NULL          |          NULL |        NULL |
+---------+---------+------------------------------------------------------------------------------------------------------------------------------------------------+----------------------------------+---------+-----------------------------------------+----------+----------+--------+------------+---------------+-----------+---------------+-------------+----------+---------------+---------------+-------------+
**5 rows in set (1.10 sec)**

This is a tests database, with the following records:

mysql> select count(*) from oc_filecache;
+----------+
| count(*) |
+----------+
|  7019685 |
+----------+
1 row in set (0.28 sec)

mysql> select count(*) from oc_filecache_extended;
+----------+
| count(*) |
+----------+
|      181 |
+----------+
1 row in set (0.00 sec)

mysql> 

Im too tired to try to update the SelectFileCache function right now. But I will give it a try tomorrow... awesome finding!

**Comment by ClCfe** [comment]
Hello

`ALTER TABLE test.oc_filecache ADD INDEX 'fs_storage_name'('storage', 'name') USING BTREE;`

The index part on name will never be used because the query is name like %...%

Thus is wasting storage and slowing inserts/updates in this table

**Comment by glorenzutti** [comment]
Ok, I don't know that much php. I don't know anything about php.

I have been looking at the selectFileCache function and although it is quite understood, I can't find that it is only that function that needs to be changed.

It seems to me that a new function needs to be created to add that Where.

The functions appear to be called in the Cache.php file, and the query is built there. So with my limited understanding, it seems to me that you have to put together the query in that file, to add the Where that makes everything work.

I hadn't noticed @ClCfe's comment! You are right.

**Comment by skjnldsv** [comment]
Search in 22 received some really big improvements.

**Comment by ricardopadilha** [comment]
> I will try to modify the SelectFileCache function in /lib/private/Files/Cache/CacheQueryBuilder.php to match the query at step 3, but it may take some time and I make no promises I will succeed. If one of you succeeds faster than me, then please do.

This is the closest I could get to a patch, but unfortunately it does not work due to the apparent inability of the QueryBuilder framework to generate subselects.

```
--- var/nextcloud/lib/private/Files/Cache/Cache.php.orig	2021-07-08 11:40:09.634932218 -0300
+++ var/nextcloud/lib/private/Files/Cache/Cache.php	2021-07-08 12:12:02.677964978 -0300
@@ -784,8 +784,12 @@

 		$query = $this->getQueryBuilder();
 		$query->selectFileCache()
-			->whereStorageId()
-			->andWhere($query->expr()->iLike('name', $query->createNamedParameter($pattern)));
+			->where($query->expr()->in('filecache.fileid',
+				$query->subselect('fileid')
+				->from('filecache')
+				->whereStorageId()
+				->andWhere($query->expr()->iLike('name', $query->createNamedParameter($pattern)));
+			))

 		$result = $query->execute();
 		$files = $result->fetchAll();
```

If there is a `subselect` function (or equivalent) somewhere, please someone let us know.

**Comment by ChristophWurst** [comment]
> If there is a `subselect` function (or equivalent) somewhere, please someone let us know.

https://github.com/nextcloud/mail/blob/ed9ba94102e8596e8adc31197fa60c58338a09d7/lib/Db/MessageMapper.php#L1008-L1026

**Comment by ricardopadilha** [comment]
> `ALTER TABLE test.oc_filecache ADD INDEX 'fs_storage_name'('storage', 'name') USING BTREE;`
> 
> The index part on name will never be used because the query is name like %...%
> 
> This is wasting storage and slowing inserts/updates in this table

And both @glorenzutti and I managed to reproduce @iwrbr's numbers. That index might not be ideal, but it does certainly work.

**Comment by andreas-p** [comment]
I can imagine that full-scanning the index is moving less data blocks than scanning the row data, EXPLAIN would prove this.

**Comment by ricardopadilha** [comment]
> EXPLAIN would prove this

For what it's worth, here's the EXPLAIN for the original query:

```
EXPLAIN SELECT filecache.fileid, storage, path, path_hash, filecache.parent, name, mimetype, mimepart, size, mtime, storage_mtime, encrypted, etag, permissions, checksum, metadata_etag, creation_time, upload_time FROM oc_filecache filecache LEFT JOIN oc_filecache_extended fe ON filecache.fileid = fe.fileid WHERE (storage = 1) AND (name COLLATE utf8mb4_general_ci LIKE '%ZAM%');
+------+-------------+-----------+--------+----------------------------------------------------------------------------------------------+----------------------+---------+----------------------------+--------+-------------+
| id   | select_type | table     | type   | possible_keys                                                                                | key                  | key_len | ref                        | rows   | Extra       |
+------+-------------+-----------+--------+----------------------------------------------------------------------------------------------+----------------------+---------+----------------------------+--------+-------------+
|    1 | SIMPLE      | filecache | ref    | fs_storage_path_hash,fs_storage_mimetype,fs_storage_mimepart,fs_storage_size,fs_storage_name | fs_storage_path_hash | 8       | const                      | 636602 | Using where |
|    1 | SIMPLE      | fe        | eq_ref | PRIMARY                                                                                      | PRIMARY              | 8       | nextcloud.filecache.fileid | 1      | Using where |
+------+-------------+-----------+--------+----------------------------------------------------------------------------------------------+----------------------+---------+----------------------------+--------+-------------+
```

And here's for the new query, with the new index:
```
+------+-------------+--------------+--------+------------------------------------------------------------------------------------------------------+-----------------+---------+-------------------------------+--------+--------------------------+
| id   | select_type | table        | type   | possible_keys                                                                                        | key             | key_len | ref                           | rows   | Extra                    |
+------+-------------+--------------+--------+------------------------------------------------------------------------------------------------------+-----------------+---------+-------------------------------+--------+--------------------------+
|    1 | PRIMARY     | oc_filecache | ref    | PRIMARY,fs_storage_path_hash,fs_storage_mimetype,fs_storage_mimepart,fs_storage_size,fs_storage_name | fs_storage_name | 8       | const                         | 636476 | Using where; Using index |
|    1 | PRIMARY     | filecache    | eq_ref | PRIMARY                                                                                              | PRIMARY         | 8       | nextcloud.oc_filecache.fileid | 1      |                          |
|    1 | PRIMARY     | fe           | eq_ref | PRIMARY                                                                                              | PRIMARY         | 8       | nextcloud.oc_filecache.fileid | 1      | Using where              |
+------+-------------+--------------+--------+------------------------------------------------------------------------------------------------------+-----------------+---------+-------------------------------+--------+--------------------------+
```

My execution times are even more absurd than @iwrbr's, i.e., I have over 36m30sec for the original query, and only a couple of seconds for the second version.

**Comment by glorenzutti** [comment]
> > If there is a `subselect` function (or equivalent) somewhere, please someone let us know.
> 
> https://github.com/nextcloud/mail/blob/ed9ba94102e8596e8adc31197fa60c58338a09d7/lib/Db/MessageMapper.php#L1008-L1026

awesome. im do not have the time right know, but I will try to check this out at the weekend.

**Comment by ricardopadilha** [comment]
> https://github.com/nextcloud/mail/blob/ed9ba94102e8596e8adc31197fa60c58338a09d7/lib/Db/MessageMapper.php#L1008-L1026

Thanks for the heads-up. The patch below is as close as I got to an implementation of the SQL query from https://github.com/nextcloud/server/issues/23835#issuecomment-829717863 but I could not confirm it is generating the exact same query. Calling `$this->logger->debug('New query: ' . $query->getSQL() . '\n');` does not seem to generate in any log file that I could find.

I suspect this patch does *not* entirely solve the problem because, although sometimes the file search works almost instantly, other times I'm still seeing file search timeouts. It seems as if that this is not the only entry point for file search.

In any case, I'm leaving it here so that other people can try to build from it.

```
--- var/nextcloud/lib/private/Files/Cache/Cache.php.orig	2021-07-08 12:25:24.949799654 -0300
+++ var/nextcloud/lib/private/Files/Cache/Cache.php.new	2021-07-08 17:28:02.699430642 -0300
@@ -782,11 +782,16 @@
 			return [];
 		}

-		$query = $this->getQueryBuilder();
-		$query->selectFileCache()
+		$subquery = $this->getQueryBuilder();
+		$subquery->select('fileid')
+			->from('filecache')
 			->whereStorageId()
 			->andWhere($query->expr()->iLike('name', $query->createNamedParameter($pattern)));

+		$query = $this->getQueryBuilder();
+		$query->selectFileCache()
+			->where($query->expr()->in('filecache.fileid', $query->createFunction( '(' . $subquery->getSQL() . ')' )));
+
 		$result = $query->execute();
 		$files = $result->fetchAll();
 		$result->closeCursor();
```

**Comment by glorenzutti** [comment]
@ricardopadilha when you see that the search timeouts, you could check your running process on the mysql to check what its going on. maybe you are right, maybe its something else.

**Comment by ricardopadilha** [comment]
Is there a simple way to log the generated SQL somewhere?

**Comment by andreas-p** [comment]
I just tested nc22 on a small storage. I'm not impressed... The user experience feels chewy.

**Comment by glorenzutti** [comment]
> Is there a simple way to log the generated SQL somewhere?

If there is, I couldn't find where.
But the query takes a long time, and it gives you the opportunity to see it by doing a "show running process" on the mysql server.

**Comment by andrimont** [comment]
Hello @all 
I am not sure this this is the right thread, let me know in such a case.
I tried to use the unified search and it break NC. I did upgrade from NC20 to NC 21 and still a search breaks the site.

1. Search is endless
2. At best it show only emails…
3. When reload the site, it breaks : **502 Bad Gateway // nginx**
4. ![Screenshot 2021-07-12 at 12 04 11](https://user-images.githubusercontent.com/19648093/125269720-46fb0980-e309-11eb-91a7-3b027edffee2.png)

> How can I check a log on this kind of issue ?
> Can I improve the search or chose a search only files (e.g. not the endless email) ?
Thank's for help.

**Comment by glorenzutti** [comment]
> Hello @ALL
> I am not sure this this is the right thread, let me know in such a case.
> I tried to use the unified search and it break NC. I did upgrade from NC20 to NC 21 and still a search breaks the site.
> 
>     1. Search is endless
> 
>     2. At best it show only emails…
> 
>     3. When reload the site, it breaks : **502 Bad Gateway // nginx**
> 
>     4. I have to reboot the site server…
>        ![Screenshot 2021-07-12 at 12 04 11](https://user-images.githubusercontent.com/19648093/125269720-46fb0980-e309-11eb-91a7-3b027edffee2.png)
> 
> 
> > How can I check a log on this kind of issue ?
> > Can I improve the search or chose a search only files (e.g. not the endless email) ?
> > Thank's for help.

Check you running process on the sql server. You may be running out of available connections. You can kill them, or restart the sql server to temporary fix this.

**Comment by dsisysteme** [comment]
Hi @all,

We just upgrade Nextcloud from v19 to v22 and have the same issue. The unified search doesn't work, and the SQL request get "stuck" in MariaDB:

`| 18244 | nextcloud-test | dsi-sys-tfoiche.univ-lille.fr:42636    | nextcloud-test | Query   |    9 | Sending data             | SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `metadata_etag`, `creation_time`, `upload_time` FROM `oc_filecache` `file` LEFT JOIN `oc_filecache_extended` `fe` ON `file`.`fileid` = `fe`.`fileid` WHERE (`storage` = 3193) AND (((`path`  COLLATE utf8mb4_general_ci LIKE 'files/RH/Maurice Dupont/%') OR (`path_hash` = 'b81a9638e4deb741e8e9c15738ee78801')) AND (`name`  COLLATE utf8mb4_general_ci LIKE '%ee%')) ORDER BY `mtime` desc LIMIT 5 |    0.000 |`

and the DB server will crash after quite some time due to overloading.....

We try to repoduce that with DBeaver and we realize that it's work when we change the ORDER BY by "storage_mtime"....

SELECT 
	`file`.`fileid`, 
	`storage`, 
	`path`, 
	`path_hash`, 
	`file`.`parent`, 
	`name`, 
	`mimetype`, 
	`mimepart`, 
	`size`, 
	`mtime`, 
	`storage_mtime`, 
	`encrypted`, 
	`etag`, 
	`permissions`, 
	`checksum`, 
	`metadata_etag`, 
	`creation_time`, 
	`upload_time` 
FROM `oc_filecache` `file` 
LEFT JOIN `oc_filecache_extended` `fe` 
ON `file`.`fileid` = `fe`.`fileid` 
WHERE (`storage` = 3193) 
AND (((`path`  COLLATE utf8mb4_general_ci LIKE 'files/RH/Maurice Dupont/%') 
OR (`path_hash` = 'b81a9638e4deb741e8e9c15738ee78801')) 
AND (`name`  COLLATE utf8mb4_general_ci LIKE '%2020%')) 
ORDER BY "mtime" desc 
LIMIT 5

For information, the number of lines in the oc_filecache table:

+--------------+
| row_count |
+--------------+
|  41101925 |
+--------------+

Maybe it could help...

**Comment by ohthehugemanatee** [comment]
> Is there a simple way to log the generated SQL somewhere?

Yes, use MySQL/MariaDB's "Slow query log" feature. 

- https://dev.mysql.com/doc/refman/5.7/en/slow-query-log.html
- https://mariadb.com/kb/en/slow-query-log-overview/

**Comment by gennaios** [comment]
Has anyone considered full-text search? Think of search that we use elsewhere: the forums, desktop search (file manager, calendar, other), GitHub issues itself, the browser address bar, etc. Phrase search makes it very difficult to find results.

I think ideally file search would allow one to use any number of search terms in any order plus include full path where folder names could be indicative of contents. FTS would be useful in other places too such as calendar and contacts where terms could occur anywhere in multiple fields (name, title, location, city, description, etc.). 

It could be a fair amount of effort to support all supported databases, but over the long term, I think it is the best solution.

**Comment by glorenzutti** [comment]
> Has anyone considered full-text search? Think of search that we use elsewhere: the forums, desktop search (file manager, calendar, other), GitHub issues itself, the browser address bar, etc. Phrase search makes it very difficult to find results.
> 
> I think ideally file search would allow one to use any number of search terms in any order plus include full path where folder names could be indicative of contents. FTS would be useful in other places too such as calendar and contacts where terms could occur anywhere in multiple fields (name, title, location, city, description, etc.).
> 
> It could be a fair amount of effort to support all supported databases, but over the long term, I think it is the best solution.

Good luck with that.. I did try the fts plugin and in my environment does not work at all. I was trying to index 1300 users files, in several file servers, with millions of files and its a no go. I don't recall all the issues, but some of them are:

1) They are several issues with the code page of the filenames and the index process. The index process sometimes just crash, or hangs forever.
2) I was dedicating a server with 32GB of ram xeon silver processor to only do the index, a week later and it was 8% done, without taking notice of every change that was made... the entire week, by all the users. Its impossible to predict when its going to finish and you cant run the first index, and the monitor index at the same time.
3) The first run cant run in parallel to take advantage of modern hardware, or even some kind of horizontal scaling. If I could say to the index job "index THIS user" or "index THIS external storage" it would be AWESOME. but.. u cant...

At least for me, the FTS didn't work at all...

**Comment by gennaios** [comment]
Perhaps developers more experienced with databases will chime in.

I do not mean the FTS plugin which I believe uses Elasticsearch. I mean instead of using the current scheme of LIKE queries, refactor Nextcloud to instead use the built-in FTS capabilities of SQLite, MySQL and, PostgreSQL. Decide on the relevant metadata fields (e.g., name + path for files; name, title, address, city, etc. for contacts; title, description, … for calendar), and then create FTS indices for them.

If there are not current developers who have experience with FTS, then it may take a bit of time. But in the end, once done in the best possible way supported by each database - indices are created differently for each db and query syntax differs –, it will be significantly more performant. As said above, imagine any place where search is done across a large number of entries, FTS is the way. Think of the Amazon catalog for instance, certain fields are indexed, and while users will not have their hardware, it will be very fast. Perhaps later, like Redis is used to speed up certain things, there may be future support for further improvements with Sphinx.

Relevant pages for each database:

https://www.sqlite.org/fts3.html
https://www.sqlite.org/fts5.html
https://dev.mysql.com/doc/refman/8.0/en/fulltext-search.html
https://www.postgresql.org/docs/9.6/textsearch.html

**Comment by skjnldsv** [comment]
Like said:

> Search in 22 received some really big improvements.
> Testing a simple search amongst 10k files on my dev setup:
> 
> ```
> 21: 1660ms
> 22:  506ms
> ```

23 will also see some performances improvements.

**Comment by gennaios** [comment]
Perhaps someday more will request such and understand its importance. I think eventually FTS will be implemented. LIKE queries with exact phrase doesn't work well as the data set grows. Imagine any search we constantly use such as browser bookmarks and history. Any number of search terms specified in any order found across multiple columns, e.g. name path title description place etc. I think is the only way. The main resistance is time and effort. 

I only have a small amount of experience with FTS on Mysql and Pgsql, yet I have done it at least once. Ok experience with Sqlite. If someone someday wants to implement such, perhaps create a new issue and mention me. If I have the time, I might be able to help develop the needed SQL. If someone decides eventually to start and implement, I'll try to help here and there as best I can.

**Comment by lopezio** [comment]
HI all, I don't understand how things like this can make it to a new major release without any option to use the old method in between. File search has become **unusable** coming from 19 to 21. I'm very sad that this previously so useful feature (particularly for instances with many files) has been destroyed in such a way. :(

**Comment by skjnldsv** [comment]
@lopezio sometimes a software have to go through different phases to cleanup its code and grow from a specific point. This is still an opensource software and any help is appreciated. Nextcloud is used by millions of people with a lot of different setups, we unfortunately can't focus on all directions :disappointed:

**Comment by lopezio** [comment]
> @lopezio sometimes a software have to go through different phases to cleanup its code and grow from a specific point. This is still an opensource software and any help is appreciated. Nextcloud is used by millions of people with a lot of different setups, we unfortunately can't focus on all directions 😞

@skjnldsv I know. I appreciate FOSS and your work and that of the other authors here! Yet in a product like nextcloud, on which many people, companies, appliances, and more, rely, the missing QA _is_ a bit of a problem. I'm not sure what the real solution would be here, and I'm sure it has been discussed before. The business options from Nextcloud GmbH have an unaffordable price tag, so I can't pass them onto most of my customers. They're at least meant for bigger organizations than those I serve, and it remains to be seen if that would help in the present case.
One example would be accepting donations - like other projects  - to help focus on specific issues. The wrong thing about it is that some counterparts (e.g. appliances really making money with this) should be donating by default, instead of sitting back and waiting for the community to sort everything out. Still, that could help. I for myself would be among the first to offer money to contribute fixing such an issue. Although I'd probably have the skills to get into it myself php-wise, I'd need extra time to get into the implementation details and tests - so paying a part of it would be totally fair, when I have other things on my mind and list. If only the people who've read this thread and added thumbs did the same or similar, there would be a good start of financial room to fix this without requiring anyone to starve. And without taking everything for granted, as some people do, which is not right.

Still, to all of you who worked on the unified search project: guys/gals, how could you miss this when making such a major change? OK, don't tell me. I guess the whole covid thing started to ramp up and you also got family, kids, parents, friends, and more, and like everyone there were plenty of other much more important thoughts to follow. I get this. 

But as of now, is there any patch that was submitted to 22 that helps? Or any other unpublished patch to put into an instance (running 21 or 22) to help testing? 

Don't get me wrong: I am a big fan of this community and I am grateful for this product and the mostly awesome work you've been doing. But this here is a _major problem_, it just can't take so long to be solved, even less across 2 major releases. In some cases it does really bring down the whole instance. I've seen it happen more than once.

Very respectfully and with Best Regards. And sorry for the non-technical (so, sort of OT) post.

**Comment by lopezio** [comment]
> > https://github.com/nextcloud/mail/blob/ed9ba94102e8596e8adc31197fa60c58338a09d7/lib/Db/MessageMapper.php#L1008-L1026

FYI: Tested this patch on my 21.0.13 installation. Patch went through, but no change. Also after adding the index mentioned above. Search still unusable...
Also keep in mind that with a properly set up database query results will be cached. So issuing the same query (without any change) twice will easilly yield a 10x performance benefit - but once you change the argument in the LIKE it shows the "real" time again.

**Comment by lopezio** [comment]
> Like said:
> 
> > Search in 22 received some really big improvements.
> > Testing a simple search amongst 10k files on my dev setup:
> > ```
> > 21: 1660ms
> > 22:  506ms
> > ```
> 
> 23 will also see some performances improvements.

@skjnldsv I can confirm that 21.1.0 makes a difference, a notable one. It's still far from the performance I was used on 19.x, but it is at least back to usable, until now without the breakages I've seen before.

**Comment by ricardopadilha** [comment]
> Nextcloud is used by millions of people with a lot of different setups, we unfortunately can't focus on all directions

I'm sorry to be blunt, but this is not just any direction. This is a DoS-spawning, production-breaking, only-a-restart-can-fix bug. If you are lucky, search stops working. If you are not, then your DB server is completely frozen with queries that take several (dozens of) minutes.

I understand that resources are limited and that issues have to be prioritized -- I have to do that kind of prioritization every day at my job. What is disappointing to me is the priority of this particular issue. From the beginning, the impact of this has been downplayed, and hand-waved away until @iwrbr's findings.

I guess what everyone is asking for here is some kind of commitment that this will be prioritized correctly, and therefore that dev time will be assigned accordingly. In fact, just as @lopezio mentioned, people (including me) are willing to pay to have the priority on this bumped up.

I've tried to look into the code (see my previous comments), but I have enough coding experience to know that changing that part of the code requires a level of familiarity with the codebase that is not achieved by simply reading a couple of source files. That is work that needs to be done by whoever refactored that code recently, because they are likely the only one person who understands all the potential ramifications and side effects.

**Comment by lopezio** [comment]
> I'm sorry to be blunt, but this is not just any direction. This is a DoS-spawning, production-breaking, only-a-restart-can-fix bug. If you are lucky, search stops working. If you are not, then your DB server is completely frozen with queries that take several (dozens of) minutes.

@ricardopadilha are you still experiencing the DoS-spawning and extreme slowness with 22.1.0? I agree with all you say, and as a fun fact, when upgrading nextcloud instances lately, I became more conservative and started avoiding "x.0.0" versions. That's why I didn't upgrade to 22 but 21 in the first placee. Now that 22.1.0 is out, I gave it a try, and the breakage seems to be fixed on the instances I am handling. Although the way search works is still slower and less efficient (and buggy: sometimes I click on a folder in the search results and it opens something completely different. But that'd be something for a new ticket...).

**Comment by andreas-p** [comment]
The main issue for me wasn't the single full table scan query, but observing multiple identical queries: each keypress multiplied by number of shares. While I have just 4 shares, other users have 20-40, making it brutally worse.
I could live with a single query on each keypress (while still not amused). Is this resolved on NC22?

**Comment by glorenzutti** [comment]
Before the last update of nextcloud, it happened to me a lot that the database server crashed due to DOS.
With the latest version I have had DOS very occasionally, with the same queries as before.
It is important to clarify that I have 14 million files indexed in multiple external storages.

**Comment by ricardopadilha** [comment]
> are you still experiencing the DoS-spawning and extreme slowness with 22.1.0?

Unfortunately, I'm stuck on 21 until all the apps that I need are updated for 22 (in particular group folders: https://apps.nextcloud.com/apps/groupfolders), so I don't know if 22 is better in this regard or not.

**Comment by lopezio** [comment]
> The main issue for me wasn't the single full table scan query, but observing multiple identical queries: each keypress multiplied by number of shares. While I have just 4 shares, other users have 20-40, making it brutally worse.
> I could live with a single query on each keypress (while still not amused). Is this resolved on NC22?

@glorenzutti It looks better on 22.1.0 when looking at the processlist than before (but I need to dig deeper to be sure...), and also, it doesn't trigger any queries with only one char. In general they are much faster. Haven't seen the DoS yet on an instance with about about 2.5 million oc_filecache records. Please note that while I have yet to find in the nc changelog something that hints at this issue, 22.**1**.0 brought so many bugfixes that I'm surre we should be differenciating here between 22.**0**.0 and 22.**1**.0 .

**Comment by lopezio** [comment]
> The main issue for me wasn't the single full table scan query
> Is this resolved on NC22?

@andreas-p Looking at:
https://github.com/nextcloud/server/commits/v22.1.0/lib/private/Files/Cache - it looks like the single query made it to 22.1.0 thanks to @icewind1991 commit here: https://github.com/nextcloud/server/commit/5d1d53c42e9a8342f9fd2da815e6653a64203c0c#diff-d4b7cb2812a4e30c1e474f2fd251076344a9e042a708f6b830f55ca786226b69
There are also a bunch of other commits there pertaining search - let's hope they will be backported to a 21.0.5 or such..

**Comment by andreas-p** [comment]
@lopezio Thanks for the pointer, I'll update to NC22 and will report back my findings.

**Comment by andreas-p** [comment]
Testing with 22.1, I can confirm that the multiple database query issue was resolved; only one query per keypress (still too much when typing rapidly).
The UI is still inferior, and results can be very irritating (only 3 files found with 3 chars entered, but more with 4 chars), but that's another issue.

**Comment by PVince81** [comment]
the fix for the multiple queries is here: https://github.com/nextcloud/server/pull/26874

before that, the problem was that the more received shares you have (mount points), the more search queries it would do, it would basically run the same search in each of the mount point.

now on NC 22.1 after the fix, the number is reduced to a single one which combines the storages there

**Comment by gennaios** [comment]
@dmuensterer do you think path should also be included in the index? As others may not have been used to FTS, they may not see all the benefits and possibilities. Imagine searching any catalog like Amazon or using any search on a large collection. Phrase search becomes less effective.

For example, I store my Calibre library on Nextcloud, a large number of epubs. With folders such as:

>../Calibre Library/William Shakespeare/Hamlet (4784)/Hamlet - William Shakespeare.epub
> ../Calibre Library/Yves Bonnefoy/L'Hesitation d'Hamlet et la Decision de Shakespeare (13796)/L'Hesitation d'Hamlet et la Decision de Sh - Yves Bonnefoy.epub

It would be nice to be able to use any number of search terms specified in any order. I think it is preferable to also append the wildcard operator to each word, e.g. shakespea* would find all words that start with. There is also ordering by relevance. From a quick look at the docs, and trying some queries, this is what I have so far:

`SELECT …, MATCH(name, path) AGAINST ('+shakespear* +hamlet*' in BOOLEAN MODE) AS score from oc_filecache where  MATCH(name, path) AGAINST ('+shakespear* +hamlet*' in BOOLEAN MODE) order by score desc;
`
If such is done, perhaps there needs to be some pre-parsing of the query string, such that a plus is added before each term if it doesn't exist, a minus is left alone, etc. Such operators may also differ for the different dbs.

**Comment by dmuensterer** [comment]
Hmm, @jancborchardt how many files did you have when experiencing the delay.
I ran a benchmark myself and couldn't observe slowness like this.
I created 100k testfiles and any search query in the database takes around 0.02s-0.03s.
I.e. theoretically this should scale linearly due to O(n) so in order to be in a range of ~40secs this would be 20 Million files per user.
@gennaios I think the absolute best way for this would be using wildcards for each word.
Ordering by relevance is IMO not needed as this is a file search, not a search to find multiple pages containing very similar content. Maybe still something to look for in the future?

I'll create a couple million files as a benchmark and see where that puts me performance wise when searching.
Maybe things habe been improved already since @jancborchardt created the issue?

**Comment by icewind1991** [comment]
A proper FTS index hasn't been implemented yet mainly because it request database specific implementations for both creating the index and generating the query.

**Comment by biva** [comment]
@dmuensterer We have 15000 files and 2000 links, the file search is really slow as well (5 seconds)

**Comment by DLalot** [comment]
Nothing better for us in 22.2.0
oc_filecache has 83 367 882 rows
Galera Cluster mariadb
We have a cron to kill SQL requests longer than 60 sec. We have seen search requests lasting more than 1 hour :-(
Search is no more working from NC19

**Comment by acsfer** [comment]
Here is a single search, single keyword, on v22.2.0.
MariaDB 10.4, 16 cores, 32GB RAM box (all dataset fits in `innodb_buffer_pool`).

**27 minutes**, one result returned (expected).
_(even if nobody waits 27mn for a search result...)_

```
> select count(*) from oc_filecache;
+----------+
| count(*) |
+----------+
| 12.447.556 |
+----------+
1 row in set (3.575 sec)
```

![image](https://user-images.githubusercontent.com/12234510/140891092-c6960215-3c3f-4272-bf08-4f4989921077.png)

**Comment by icewind1991** [comment]
https://github.com/nextcloud/server/pull/29300 should help mysql to not pick the worst index to use

**Comment by andreas-p** [comment]
Trying to tune some specific query planner isn't going to fix this issue; full table scanning is a fully broken approach. 

#29614 looks promising, maybe this issue should be closed in favor of that.

**Comment by DLalot** [comment]
Hi,
Using 22.2.3, it's a little bit better as you get quickly files in the current directory, but after it's also spending time. So the best way would be not to create directories :-)
This request is ending with an error as we kill too long requests
https://xxx/ocs/v2.php/search/providers/files/search?term=annexe&from=%2Fapps%2Ffiles%2F%3Fdir%3D%2F%26fileid%3D157402006
I don't have the corresponding SQL request, but I suppose that nothing have changed. My colleague @gregamu  proposed to use a force index

**Comment by dmuensterer** [comment]
I know this is not a satisfying solution, but here’s my take on this:
Any nextcloud instance with a lot of files runs only well when using ElasticSearch instead of the built in Nextcloud Unified Search. 
When unified search takes 1minute to complete a search, Elastic takes <0.5s.
It is absolutely crazy how bad the Unified Search is compared to what’s possible.
This obviously requires a Elastic Instance running along with Nextcloud and requires the following three Nextcloud Apps:
- Full Text Search
- Full Text Search Elastic
- Full Text Search Files

This works flawlessly with n amount of files in my experience.

**Comment by glorenzutti** [comment]
> I know this is not a satisfying solution, but here’s my take on this: Any nextcloud instance with a lot of files runs only well when using ElasticSearch instead of the built in Nextcloud Unified Search. When unified search takes 1minute to complete a search, Elastic takes <0.5s. It is absolutely crazy how bad the Unified Search is compared to what’s possible. This obviously requires a Elastic Instance running along with Nextcloud and requires the following three Nextcloud Apps:
> 
>     * Full Text Search
> 
>     * Full Text Search Elastic
> 
>     * Full Text Search Files
> 
> 
> This works flawlessly with n amount of files in my experience.

Sadly, that doesn't work when you have millions of files. elastisearch never ever finishes indexing everything. I have been more than two weeks waiting for it to end.

The nextcloud detects that when you share a share with username and password among N users, it is the same share and the same files. The elastisearch fails and scans everything back.

**Comment by dmuensterer** [comment]
Interesting. I didn’t have that problem yet.
My installation has around 1.2 million files (barely shared with other users), and indexing took around 35minutes.
In your case I’d try using an Elastic Cluster with at least 3 nodes and having a go for indexing again. I’d rather spend two weeks initially indexing than waiting a minute for each search in the future…

**Comment by andreas-p** [comment]
Actually, Full Text Search appears off-topic. It has nothing to do with locating a file by its name within a specified well-known context. Using Elastic for that is like using a sledge hammer to crack a nut.

**Comment by dmuensterer** [comment]
Well it doesn’t seem like elastic is a sledgehammer when looking at the problems with the built in search. You can use elastic simply to search for files and not index their content.
I agree though that it’s off topic and there should be a scalable internal solution to this.

**Comment by dmuensterer** [comment]
I can confirm that ElasticSearch 7.16.3 with Nextcloud 23 DOES NOT index shared files multiple times for each user.
I created a shared volume while monitoring the index.

**Comment by gennaios** [comment]
@dmuensterer, As you previously experimented with using the native FTS capabilities of MySQL/MariaDB, any reason you're not continuing to use such? I am of the strong view that eventually the as one might say the "right" way is to use the built-in FTS features of each DB; all supported DBs feature such. As @icewind1991 commented such has not yet been done since it requires different queries and indexing for each DB, yet I really don't see any other option. The native FTS of each DB really is the most performant and it is specifically meant for such situations. Perhaps an option is just to not support SQLite FTS, use Sphinx, and say for performant search, one needs to use MySQL or PostgreSQL.

**Comment by andreas-p** [comment]
@gennaios FTS is NOT the solution to searching filenames. Filenames don't have a language.

Using a good query is the solution, as easy as
`Select fileid FROM oc_filecache WHERE parent=:currentDirId AND name LIKE :whatchaSearchingFor`

**Comment by gennaios** [comment]
FTS as provided by SQLite, MySQL, and PostreSQL is not like ElasticSearch as you might think. Please look into it. One main advantage is being able to search a file name, calendar event, etc. by any number of wild card terms specifying them in any order. Imagine any catalog search, if one could only search by exact phrase, really how useful would that be? FTS in that sense is used more or less in all such cases.

**Comment by dmuensterer** [comment]
> @gennaios FTS is NOT the solution to searching filenames. Filenames don't have a language.
> 
> Using a good query is the solution, as easy as `Select fileid FROM oc_filecache WHERE parent=:currentDirId AND name LIKE :whatchaSearchingFor`

This is certainly not a “good query” to find file names. This provides O(n) which is really bad. A large directory will still take minutes to be searched.

**Comment by andreas-p** [comment]
@dmuensterer In practice, completely OT. If a directory needs minutes to search it would have to contain some 10^9 files or so.

**Comment by gennaios** [comment]
Some are confusing FTS as implemented by some software that indexes file contents itself with FTS as a concept. 

FTS is the only way. Indexing file name, plus folder full path, perhaps owner or other if relevant, and for other content types such as contact and calendar name addresses metadata. 

The FTS as provided by Sqlite MySQL and Postgrsql is designed for this purpose. Data should be able to grow to any size, millions of rows, and search should still be near instant. 

Even try using your music or contact apps on your phone. Search terms can be specified in any order. Does one need exact phase like first last or last first or exact phrase in some song album or artist name? No. How can one even remember such as one’s data size grows? Does no one have some file from years ago that they might be able to find by some combination of terms but cannot recall some exact phrase? Etc.

**Comment by gennaios** [comment]
This topic really isn’t going anywhere. I’m not sure how to say such without being offensive. It seems that the majority of those that work on and contribute to Nextcloud just do not have that much experience with databases. If such were not the case, someone would have before instantly known what is the right, most commonly used path, used by others in such situations, and would have just done it. Perhaps this will continue with endless debate, discussion, etc., well meaning solutions by those just do not have enough experience and think their suggestion is the way.

So far it seems only one contributor, has stated why FTS has not been implemented:

> … because it request [sic] (requires) database specific implementations for both creating the index and generating the query.

Whenever someone gets around to it, once it’s done, it will be obvious to anyone that uses it. Any combination of joins will allow one to filter by whatever one wants: current folder, date range, owner, share status, etc., one will be able to easily add or remove filter criteria, resort by relevance/best match, date, or other, and such will be as fast as possible. Databases will be able to grow to any size, support any size organization, and it will continue to work as it should. Someday too perhaps, just as further optimization is possible with Redis, etc., Sphinx, and perhaps other, will be a configurable option for better performance.

The way discussion is going, it is as if all have no awareness of an entire world of apps and databases, and what is used elsewhere.

Someday.

**Comment by glorenzutti** [comment]
> I can confirm that ElasticSearch 7.16.3 with Nextcloud 23 DOES NOT index shared files multiple times for each user. I created a shared volume while monitoring the index.

I will try to upgrade NC to 23 this weekend and check this out. Thanks for your comment.

**Comment by andreas-p** [comment]
> The FTS as provided by Sqlite MySQL and Postgrsql is designed for this purpose. Data should be able to grow to any size, millions or billions of rows, and search should still be near instant.

It is not. A file name is not a document that can be split into lexemes in a meaningful way, let alone a Full Text. And it doesn't have a language. In a multilanguage environment, you won't get happy.

OTOH, some fine Nextcloud programmer invented the index fs_parent_name_hash, representing the optimal access method for a filename in a directory.

**Comment by dmuensterer** [comment]
If FTS is the designed for the purpose is not the matter of if it should be used.
You can clearly see the interest behind this by the amount of feedback regarding the search.
There is a reason why the Unified Search is so slow and you keep telling us the performance is great.

When searching for files, the user typically does not know what EXACTLY he’s searching for. I.e. he’s searching for keywords or fractions inside the filename.
If the filename does not start with the search term, no index can be used as a wildcard has to be used before and after the search term.

It would be much more efficient to split filesnames using FTS by the following characters to be able to use indices when searching: space, dash, underscore, dot, comma

It is what every large cloud except Nextcloud does.
And as long as the mindset of improving doesn’t change, I’m not willing to contribute with a pull request that is clearly going to be denied.

**Comment by gennaios** [comment]
As said this discussion really isn't going anywhere. 

Mentioning such to those most expert in databases with experience of managing scalability of large datasets of millions or billions of rows wouldn't even be necessary. 

Read the book High Peformance MySQL. Read or be someone that has for years read the Percona blog. 

Read links such as this:

https://www.percona.com/blog/2018/04/04/pattern-matching-queries-vs-fulltext-indexes/

https://www.digitalocean.com/community/tutorials/how-to-improve-database-searches-with-full-text-search-in-mysql-5-6-on-ubuntu-16-04

@dmuensterer Perhaps we should just do it ourselves. I don't really know PHP but I have before once implemented FTS in in both MySQL and PostgreSQL. It's been a while but I can probably figure it out again. My only somewhat decent and recent experience is with SQLite. Various DBs of millions of rows and very fast search. I can at least provide the SQL if you can assist with the PHP. Some advanced search page such that with some button on the search results control where one could see all results as full page with possibly more options such as filters and sort, that too would be nice eventually if you are able to start on such. Overall search really needs a lot of work to be able to easily find results in a large deployment. Thousands or more contacts, events, files, etc., for such the experience could be much more improved. It is as if no one really uses search everywhere like on mobile devices, catalogs, other websites and then realizes aspects of such experience really needs to be added here.

**Comment by pjft** [comment]
Subscribing here. I can confirm that search in 19 was almost instant, and now it takes dozens of seconds at times, which is suboptimal. I did think increasing MySQL memory and cache size would make a difference, but alas it didn't.

**Comment by ohthehugemanatee** [comment]
@gennaios if you can provide the SQL it will help a lot. I or another php-head can figure out how to work that into the existing PHP...

**Comment by gennaios** [comment]
@ohthehugemanatee I might very possibly be able to sometime though I may not be able to get to it for a little while.

If you or someone could provide the current SQL for record insertion and selection, I'd have an idea of the relevant needed columns and table structure.

SQLite likely could use an external content table. In such a case, one needs manaually manage: an additional insert into the FTS table, and with any change, first grab old value before change (such as rename), provide such values for a delete statement, then insert new. Or for delete, merely grab value from file table and delete from FTS before deleteing from file. MySQL and PostgreSQL I believe are easier.

**Comment by glorenzutti** [comment]
> I'll just leave this here.. Before ![image](https://user-images.githubusercontent.com/55892859/161133093-fe6252a1-bbee-4a69-a650-79b7592ee39e.png) Query time over 1h
> 
> Add: create index t_oc_filecache_storage_mtime_name on oc_filecache (storage,mtime DESC,name);
> 
> After: ![image](https://user-images.githubusercontent.com/55892859/161133761-31064b35-6bce-4607-8861-b70e8d7282a0.png) Query time ~3s
> 
> Using index on mtime with %% on name - very bad idea. In order to prevent this, it is worth change `ORDER BY mtime` to `ORDER BY mtime+0`

Hi there! Nextcloud version?

**Comment by ustuzhanin** [comment]
21.0.9
But i`m check code. Didn't see any changes

**Comment by glorenzutti** [comment]
Tnxs! Im gonna check later.
Maybe there was some change on the schema.
Im on 23.0.2.

**Comment by ustuzhanin** [comment]
> Tnxs! Im gonna check later. Maybe there was some change on the schema. Im on 23.0.2.

```
| oc_filecache | CREATE TABLE `oc_filecache` (
  `fileid` bigint NOT NULL AUTO_INCREMENT,
  `storage` bigint NOT NULL DEFAULT '0',
  `path` varchar(4000) COLLATE utf8mb4_bin DEFAULT NULL,
  `path_hash` varchar(32) COLLATE utf8mb4_bin NOT NULL DEFAULT '',
  `parent` bigint NOT NULL DEFAULT '0',
  `name` varchar(250) COLLATE utf8mb4_bin DEFAULT NULL,
  `mimetype` bigint NOT NULL DEFAULT '0',
  `mimepart` bigint NOT NULL DEFAULT '0',
  `size` bigint NOT NULL DEFAULT '0',
  `mtime` bigint NOT NULL DEFAULT '0',
  `storage_mtime` bigint NOT NULL DEFAULT '0',
  `encrypted` int NOT NULL DEFAULT '0',
  `unencrypted_size` bigint NOT NULL DEFAULT '0',
  `etag` varchar(40) COLLATE utf8mb4_bin DEFAULT NULL,
  `permissions` int DEFAULT '0',
  `checksum` varchar(255) COLLATE utf8mb4_bin DEFAULT NULL,
  PRIMARY KEY (`fileid`),
  UNIQUE KEY `fs_storage_path_hash` (`storage`,`path_hash`),
  KEY `fs_parent_name_hash` (`parent`,`name`),
  KEY `fs_storage_mimetype` (`storage`,`mimetype`),
  KEY `fs_storage_mimepart` (`storage`,`mimepart`),
  KEY `fs_storage_size` (`storage`,`size`,`fileid`),
  KEY `fs_mtime` (`mtime`),
  KEY `fs_size` (`size`),
  KEY `fs_storage_path_prefix` (`storage`,`path`(64)),
  KEY `fs_id_storage_size` (`fileid`,`storage`,`size`),
  KEY `t_oc_filecache_storage_path_name_mtime` (`storage`,`mtime` DESC,`name`)
) ENGINE=InnoDB AUTO_INCREMENT=188162449 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED |
```
In my schema

**Comment by glorenzutti** [comment]
This is the diff with my running version.

$ diff my_filecache git_filecache 
25d24
<   KEY `fs_storage` (`storage`) USING BTREE,
26a26
>   KEY `fs_storage_path_prefix` (`storage`,`path`(64)),
28,29c28,29
<   KEY `fs_storage_path_prefix` (`storage`,`path`(64))
< ) ENGINE=InnoDB AUTO_INCREMENT=19907648 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED |
---
>   KEY `t_oc_filecache_storage_path_name_mtime` (`storage`,`mtime` DESC,`name`)
> ) ENGINE=InnoDB AUTO_INCREMENT=188162449 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin ROW_FORMAT=COMPRESSED |

**Comment by tacruc** [comment]
Hi I'm wondering if my Issue is related. I'm running Nextcloud 25rc2 and I have this querry:
`SELECT `file`.`fileid`, `storage`, `path`, `path_hash`, `file`.`parent`, `file`.`name`, `mimetype`, `mimepart`, `size`, `mtime`, `storage_mtime`, `encrypted`, `etag`, `permissions`, `checksum`, `unencrypted_size` FROM `oc_filecache` `file` LEFT JOIN `oc_vcategory_to_object` `tagmap` ON `file`.`fileid` = `tagmap`.`objid` LEFT JOIN `oc_systemtag_object_mapping` `systemtagmap` ON (`file`.`fileid` = `systemtagmap`.`objectid`) AND (`systemtagmap`.`objecttype` = 'files') LEFT JOIN `oc_vcategory` `tag` ON (`tagmap`.`type` = `tag`.`type`) AND (`tagmap`.`categoryid` = `tag`.`id`) AND (`tag`.`type` = 'files') AND (`tag`.`uid` = 'arne') LEFT JOIN `oc_systemtag` `systemtag` ON (`systemtag`.`id` = `systemtagmap`.`systemtagid`) AND (`systemtag`.`visibility` = '1') WHERE ((`tag`.`category` COLLATE utf8mb4_general_ci LIKE '%icons.png%') OR (`systemtag`.`name` COLLATE utf8mb4_general_ci LIKE '%icons.png%')) AND (((`storage` = 3) AND ((`path` = 'files') OR (`path` LIKE 'files/%'))) OR (`storage` = 176) OR (`storage` = 203) OR (`storage` = 214) OR (`storage` = 227) OR (`storage` = 228) OR ((`storage` = 76) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 89) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 89) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 4) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 4) AND ((`path_hash` = '96e07e5386d9345589a915047cfce3a8') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 4) AND ((`path_hash` = '5528724ec2bedea7c8542d53b11f0016') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 76) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 161) AND (`path_hash` = '58e1c3bca52a6832e842f73392a5bac4')) OR ((`storage` = 8) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED'29aba51cd209f7c261be0b4df9b13ba5') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 4) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 8) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 8) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 5) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path_hash` = '0c8570a357b65c15abc67651c1ec72bd') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 9) AND ((`path` = 'PATHREPLACED') OR (`path` LIKE 'PATHREPLACED/%'))) OR ((`storage` = 8) AND (`path_hash` = 'b03706e617445929129e92c4db9971d9'))) ORDER BY `mtime` + '0' desc LIMIT 5 `

And 5 other like this running for 19 hours now. Somewhere yesterday I searched for "icons.png".

**Comment by szaimen** [comment]
Hi, please update to at least 23.0.12 and report back if it fixes the issue. Thank you!

**Comment by pjft** [comment]
I think this is good on 25 already.

**Comment by szaimen** [comment]
Thanks for testing!

**Comment by pjft** [comment]
For what it's worth, I think it was also good on 24, as I was running it for a few weeks before 25 was stable.

Thanks for checking.

**Comment by gennaios** [comment]
For those that have thousands of files and potentially many search results with each search, displaying five at a time is not an ideal solution. Does such merely mask the issue?

**Comment by pjft** [comment]
I agree with the comment about the UI and only showing five results at a time, but it is unrelated to the performance issues or fixes. The performance issues were there when the UI was already only showing 5 results at a time, unless I'm mistaken. 

Nothing changed there. It's just the performance that improved.

**Comment by pjrobertson** [comment]
> Edit: still, I'd be happy with a way to open search results in a proper, long list format. Like it used to be before it got reduced to showing results in the search panel alone.

I second this. The current small search panel is not very user friendly. It's small and doesn't show the full names/file paths of files when you search, and is clunky to scroll through. Furthermore, if you type a search then hit enter ↩, it'll automatically open the first result, which is counter-intuitive.

It'd be great to have a full search page available. Gmail gives a good example of good search usability. Showing a popup window initially, and then another window if you hit enter.

**Comment by glorenzutti** [comment]
Can I second this too ? second seconder? or a thirder?
 
The UI has some room for improvement.
